### PR TITLE
Make `from_epoch` a private signal in the UST

### LINKED ISSUE #290: Don't reveal `fromEpoch` during user state transition
Revealing the `fromEpoch` during a UST gives an outside observer a lot of information about the users of the system.

e.g. say that 10 people transition to epoch 4

It's now epoch 10, and only 2 users have transition from epoch 4. We can conclude that 8 users from epoch 4 have not received attestations since epoch 4.

--

We can avoid revealing the `fromEpoch` during UST by building an incremental merkle tree of state tree roots/epoch tree roots of sealed epochs.

Say we have a binary incremental merkle tree with 2^32 leaves. Call this the "history tree". Each leaf of this tree is `H(state tree root, epoch tree root)`. When a user performs a UST instead of revealing an epoch tree root and state tree root they can reveal a history tree root.

This is safe because:
- Each epoch transition nullifier is globally unique
- Each non-empty state tree, and non-empty epoch tree root is globally unique
  - State tree leaves have the current epoch
  - Epoch tree leaves have the epoch key (which changes each epoch)

If we reveal a history tree root we don't need to reveal a `fromEpoch`. Instead it's proven to be a valid pair of state/epoch tree roots which the user has not already transitioned from.

--

To implement this we can add a leaf to the history tree when an epoch is sealed. If an epoch has no attestations and no users in the state tree, no leaf is added to the history tree (e.g. no one should transition from this epoch because no one participated in this epoch).

Using a history tree with 2^32 leaves should allow ~136 years of operation using 1 second epochs.

### PR DESCRIPTION
**Pull request by vimwitch**
fix(circuits): make from_epoch a private input

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Closes #290 

## What is the new behavior?

Adds a history tree for each attester. All epochs with either a signup, or an attestation, must be sealed. Epochs with only state tree changes (e.g. signups only) can be sealed without a zk proof.

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

File: packages/contracts/deploy/deploy.ts [hunk 1]
```diff
@@ -60,16 +62,26 @@ export const deployUnirep = async (
...
 
+    poseidonPath = 'poseidon-solidity/PoseidonT3.sol/PoseidonT3.json'
+    poseidonArtifacts = tryPath(poseidonPath)
+    poseidonFactory = new ethers.ContractFactory(
+        poseidonArtifacts.abi,
+        poseidonArtifacts.bytecode,
+        deployer
+    )
+    const PoseidonT3 = await poseidonFactory.deploy()
+    await PoseidonT3.deployed()
```
[the remark this answers is not in the document]
Comment (vimwitch) [reply packages/contracts/deploy/deploy.ts]: I think if there are more than 3 it makes sense it iterate, but for just 2 the current is okay

**Comment by vimwitch** [comment]
So in the current implementation `updateEpochIfNeeded` must be called for the epoch the user is transitioning from. This must happen so that the `HistoryTreeLeaf` event will be emitted and the proof can be calculated.

Once we have #350 we can update the latest history tree leaf any time the state tree is updated. This way the `updateEpochIfNeeded` function doesn't need to do the empty epoch seal logic. Anytime the state tree is changed the latest history tree leaf will be changed. If an attestation occurs the leaf will be zeroed (so the a user cannot transition with 0 rep).

When a previous epoch is sealed we can remove the latest leaf, insert the sealed leaf, then re-insert the latest leaf. We'll need to figure out how to handle the events for this.

I think for now requiring the attester to call `updateEpochIfNeeded` manually is an okay requirement. #350 should be done in the next week to fix this problem. There's also an optimistic approach to calculating the history leaf in #359. This approach will fail if a different epoch is sealed between when the proof is calculated and broadcasted.

**Review by vivianjeng — COMMENTED** [review]
https://github.com/Unirep/Unirep/blob/4ea2a43e9fb4dfe323e8a42d8f4f2347caace796/packages/contracts/contracts/Unirep.sol#L376-L377
I think we should note that it is an history tree root
and we should document it in protocol/trees

File: packages/contracts/contracts/Unirep.sol [hunk 2]
```diff
@@ -459,14 +472,33 @@ contract Unirep is IUnirep, VerifySignature {
         epoch = attesterCurrentEpoch(attesterId);
         if (epoch == attester.currentEpoch) return epoch;
 
-        // otherwise initialize the new epoch structures
+        // otherwise seal old epoch if empty
+        AttesterState storage state = attester.state[epoch];
+        uint oldEpoch = attester.currentEpoch;
```
Comment (vivianjeng) [inline packages/contracts/contracts/Unirep.sol:477]: I think it should take the `oldEpoch` state
```solidity
uint oldEpoch = attester.currentEpoch;
AttesterState storage state = attester.state[oldEpoch];
```

**Review by vivianjeng — CHANGES_REQUESTED** [review]

**Comment by vimwitch** [comment]
Closing in favor of #386

**Pull request closed without merging**